### PR TITLE
[SPARK-8977][Streaming] Defines the RateEstimator interface, and implements the ReceiverRateController

### PR DESCRIPTION
**Pull request by huitseeker**
Depends on #15.

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/42/

**Build Log**
_last 10 lines_

```
[...truncated 16 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/16/merge^{commit} # timeout=10
Checking out Revision 2e2fbad59abfff947c010f931d700d8753634a15 (refs/remotes/origin/pr/16/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 2e2fbad59abfff947c010f931d700d8753634a15
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 8a274e7bdf692c9bb4e41d924521e8b455de8169 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/42/ and message: Merged build finished.
```

Test FAILed.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * an estimate of the speed at which this stream should ingest messages,
+ * given an estimate computation from a `RateEstimator`
+ */
+@DeveloperApi
+class RateController(val streamUID: Int, rateEstimator: RateEstimator)
+  extends StreamingListener with Serializable {
+
+  protected def publish(rate: Long): Unit = ()
+
+  private def computeAndPublishExecutionContext = ExecutionContext.fromExecutorService(
```
Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: Shouldn't this be a `val`, or maybe even an `implicit val`?

**Comment by huitseeker** [comment]
unrelated SQL tests failing: https://ci.typesafe.com/job/ghprb-spark-multi-conf/label=Spark-Ora-JDK7-PV,scala_version=2.10/42/console

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala (hunk 1 above)
Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: That would pose a serializability problem.

Comment (dragos) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: `@transient` would fix it. Wouldn't the current code create a new thread on each rate computation?

Comment (huitseeker) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/RateController.scala]: Oh, yes, you're right !

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/44/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/16/merge^{commit} # timeout=10
Checking out Revision a32021467c61f16cc5b8181b85ab62503bf3852e (refs/remotes/origin/pr/16/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f a32021467c61f16cc5b8181b85ab62503bf3852e
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 758b9bc086159548838c9ff31e6d7ef146e77c21 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/44/ and message: Merged build finished.
```

Test FAILed.

**Comment by huitseeker** [comment]
https://ci.typesafe.com/job/ghprb-spark-multi-conf/label=Spark-Ora-JDK7-PV,scala_version=2.10/44/console
<< unrelated SQL

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/45/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/16/merge^{commit} # timeout=10
Checking out Revision a32021467c61f16cc5b8181b85ab62503bf3852e (refs/remotes/origin/pr/16/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f a32021467c61f16cc5b8181b85ab62503bf3852e
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 397b8b7b3e98904740df3f38d1b81b2508dc74b8 to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/45/ and message: Merged build finished.
```

Test FAILed.

**Comment by huitseeker** [comment]
https://ci.typesafe.com/job/ghprb-spark-multi-conf/label=Spark-Ora-JDK7-PV,scala_version=2.10/45/console << Unrelated SQL again

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/51/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/16/merge^{commit} # timeout=10
Checking out Revision 5abf73eb27c412e0035006aa33910e91e6e593b3 (refs/remotes/origin/pr/16/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 5abf73eb27c412e0035006aa33910e91e6e593b3
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result FAILURE
Notifying upstream projects of job completion
Setting status of 02b6c4b68024365a5bfbd27f61bde6a33a4b6c4f to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/51/ and message: Merged build finished.
```

Test FAILed.

**Comment by typesafe-tools** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.typesafe.com/job/ghprb-spark-multi-conf/55/

**Build Log**
_last 10 lines_

```
[...truncated 11 lines...]
 > git rev-parse refs/remotes/origin/origin/pr/16/merge^{commit} # timeout=10
Checking out Revision 5abf73eb27c412e0035006aa33910e91e6e593b3 (refs/remotes/origin/pr/16/merge)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 5abf73eb27c412e0035006aa33910e91e6e593b3
First time build. Skipping changelog.
Triggering ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10
ghprb-spark-multi-conf » Spark-Ora-JDK7-PV,2.10 completed with result ABORTED
Notifying upstream projects of job completion
Setting status of f3d7ac96e036777bed252a064113721f3074409e to FAILURE with url http://ci.typesafe.com/job/ghprb-spark-multi-conf/55/ and message: Merged build finished.
```

Test FAILed.

**Pull request closed without merging**